### PR TITLE
storage,colfetcher: implement local fast-path for COL_BATCH_RESPONSE

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit implements the local fast-path for the COL_BATCH_RESPONSE
scan format. The idea is that if a Scan request is evaluated locally
(i.e. on the same node for single-tenant deployments or within the
shared process for multi-tenant deployments), then we can avoid
the redundant serialization of the columnar batches in the Apache
Arrow format and just pass the batches as a slice of pointers through
the protobuf. Additionally, this also allows us to avoid a copy of the
data from `ScanResponse.BatchResponse` into the columnar batch.

To achieve this the ScanResponses and the ReverseScanResponses now
contain a new custom `ColBatches` message which only includes
`[]coldata.Batch` that is not marshalled as part of the protobuf
serialization.

Now that we can have a single multi-range request result in locally- and
remotely-executed single-range requests, we need to be careful when
combining them. In particular, in order to preserve the ordering between
single-range requests we now always deserialize the remotely-executed
ones (since this "combining" happens on the KV client side and won't be
sent over the wire again) while "merging" them accordingly. This
required introduction of an injected helper for the deserialization from
the Apache Arrow format into the `kvpb` package. This deserialization
also required that we have access to the `fetchpb.IndexFetchSpec` proto
that is stored in the BatchRequest, thus, the signature of `combine`
method has been adjusted to include the reference to the BatchRequest.

Additional quirk of this commit is that the `cFetcher` cannot reuse the
same batch when it is used by the `cFetcherWrapper` when skipping the
serialization. (If it did reuse batches, then the slice of batches would
contain multiple references to the same batch, so only the last reference
would be correct - all previous ones would have been reset.) To do that
the `colmem.SetAccountingHelper` has been adjusted to be able to keep
the same heuristic when it comes to the sizing of the batch while always
allocating a new one, even if under other circumstances it would have
reused the old batch.

It's also worth noting the story about memory accounting of these local
batches. The `SetAccountingHelper` used by the `cFetcher` always tracks
the memory usage only of the last batch, so we need to account for all
other batches ourselves. We go around this by providing the `cFetcher`
with a "detached" memory account (i.e. an account that is not connected
to the memory accounting system) that is used by the `cFetcher` to limit
the batch size based on the footprint, and modifying the
`cFetcherWrapper` to perform the accounting against the proper memory
account. This commit also clarifies the contract of
`CFetcherWrapper.NextBatch` that it is the wrapper's responsibility to
perform memory accounting of all batches, regardless of the return
format, against the provided memory account.

This only covers part of the story from the KV server side. On the KV
client side the memory accounting is done in `txnKVFetcher`. When the
batches are serialized, they are included in
`ScanResponse.BatchResponse` field and, thus, are included into
`BatchResponse.Size` which we use for accounting. For the non-serialized
batches this commit implements the custom `Size()` method so that the
true footprint of all `coldata.Batch`es is included into
`BatchResponse.Size`. As a result, all local batches (including the ones
that were deserialized when combining responses to locally- and
remotely-executed requests) are tracked by the `txnKVFetcher` until a
new `BatchRequest` is issued, so the ColBatchDirectScan doesn't need to
perform the accounting. (Note that we perform the accounting for
`ScanResponse.BatchResponse` slices in a similar manner - we don't
shrink the memory account when a single response becomes garbage (due
to likely under-accounting in other places).)

A special note on type schemas with enums: since enums require type
hydration that is not easily available on the KV server side and we
treat them simply as bytes values, the presence of enums forces us to
serialize the batches even for locally-executed requests. This seems
like a minor limitation in comparison to not supporting enums at all.

Another note on the datum-backed vectors: since the `cFetcherWrapper`
also doesn't have access to a valid `eval.Context`, the datum-backed
vectors produced by the wrapper are "incomplete". Previously, since we
always serialized the batches, it wasn't an issue. However, now if we
get a non-serialized batch from the locally-executed request, we must
update all datum-backed vectors with the proper eval context. This is
done by the `ColBatchDirectScan`.

The microbenchmarks of this change when the direct columnar scans are
always enabled are [here](https://gist.github.com/yuzefovich/a9b28669f35ff658b2e89ed7b1d43e38).
Note that there are three distinct operation modes in that gist:
- `Cockroach` and `MultinodeCockroach` - single-tenant deployments
- `SharedProcessTenant` - this is how we imagine that dedicated clusters
will run once the Unified Architecture is achieved
- `SepProcessTenant` - this is how we run Serverless.

For the first two this commit results mostly in a minor improvement in
latency and sometimes noticeable reducation in allocations, as expected.
SepProcessTenant config - which cannot take advantage of the local
fastpath - sees a minor slowdown in latency and no changes in
allocations, as expected (I'm attributing this to increased overhead of
the direct columnar scans and increased size of `ScanResponse` objects).

However, these are micro-benchmarks, and they don't show the full
picture. In particular, they don't process enough data and often select
all columns in the table for this feature to show its benefits. I'm more
excited about the results on the TPC-H queries. Here is the impact of
this commit on 3 node cluster running in single-tenant model (averaged
over 10 runs):
```
Q1:	before: 4.46s	after: 4.23s	 -5.15%
Q2:	before: 3.18s	after: 3.30s	 3.45%
Q3:	before: 2.43s	after: 2.11s	 -13.20%
Q4:	before: 1.83s	after: 1.84s	 0.44%
Q5:	before: 2.65s	after: 2.48s	 -6.34%
Q6:	before: 7.59s	after: 7.46s	 -1.65%
Q7:	before: 5.56s	after: 5.72s	 2.71%
Q8:	before: 1.14s	after: 1.11s	 -2.29%
Q9:	before: 5.77s	after: 5.31s	 -7.86%
Q10:	before: 1.98s	after: 1.94s	 -1.92%
Q11:	before: 0.73s	after: 0.69s	 -5.52%
Q12:	before: 7.18s	after: 6.91s	 -3.79%
Q13:	before: 1.24s	after: 1.24s	 0.16%
Q14:	before: 0.70s	after: 0.66s	 -5.32%
Q15:	before: 3.99s	after: 3.64s	 -8.89%
Q16:	before: 0.95s	after: 0.94s	 -1.16%
Q17:	before: 0.27s	after: 0.26s	 -5.49%
Q18:	before: 2.67s	after: 2.15s	 -19.39%
Q19:	before: 4.03s	after: 2.96s	 -26.46%
Q20:	before: 12.91s	after: 11.49s	 -10.98%
Q21:	before: 7.14s	after: 6.99s	 -2.13%
Q22:	before: 0.60s	after: 0.57s	 -5.48%
```

Furthermore, here is the comparison of the direct columnar scans
disabled vs enabled:
```
Q1:	before: 4.36s	after: 4.23s	 -2.91%
Q2:	before: 3.57s	after: 3.30s	 -7.63%
Q3:	before: 2.31s	after: 2.11s	 -8.61%
Q4:	before: 1.88s	after: 1.84s	 -2.07%
Q5:	before: 2.55s	after: 2.48s	 -2.70%
Q6:	before: 7.94s	after: 7.46s	 -6.04%
Q7:	before: 5.87s	after: 5.72s	 -2.61%
Q8:	before: 1.12s	after: 1.11s	 -1.07%
Q9:	before: 5.79s	after: 5.31s	 -8.27%
Q10:	before: 1.97s	after: 1.94s	 -1.47%
Q11:	before: 0.69s	after: 0.69s	 -0.29%
Q12:	before: 6.99s	after: 6.91s	 -1.16%
Q13:	before: 1.24s	after: 1.24s	 -0.48%
Q14:	before: 0.68s	after: 0.66s	 -3.37%
Q15:	before: 3.72s	after: 3.64s	 -2.23%
Q16:	before: 0.96s	after: 0.94s	 -1.88%
Q17:	before: 0.28s	after: 0.26s	 -6.18%
Q18:	before: 2.47s	after: 2.15s	 -12.87%
Q19:	before: 3.20s	after: 2.96s	 -7.35%
Q20:	before: 11.71s	after: 11.49s	 -1.88%
Q21:	before: 7.00s	after: 6.99s	 -0.06%
Q22:	before: 0.58s	after: 0.57s	 -2.07%
```
In other words, on TPC-H queries it is now already beneficial to
enable the direct columnar scans in single-tenant world (and I think
there are more minor optimizations ahead). For reference, [here](https://gist.github.com/yuzefovich/0afce5c0692713cf28712f076bab415b)
is the comparison of direct columnar scans disabled vs enabled on this
commit. It also shows that we're not that far off from reaching the
performance parity in micro-benchmarks.

Addresses: #82323.
Informs: #87610.

Epic: CRDB-14837

Release note: None

### CODE REVIEW

**Review by michae2 — COMMENTED** [review]
> Reviewed 19 of 38 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95033)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @sumeerbhola, and @yuzefovich)

___
*[`pkg/kv/kvpb/api.proto` line 645 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUYXQODdj-1ZkhUuzr:-NPUYXQODdj-1ZkhUuzs:b-oc0s5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/kv/kvpb/api.proto#L645)):*
> ```protobuf
>   //
>   // If set, rows and batch_responses will not be set.
>   // TODO(feedback wanted): should this be nullable? This will decrease the size
> ```

useless blathering: On one hand, I would expect the majority of ScanRequests to cross the network and need serialization, suggesting that we should optimize for that more common remote case and set this nullable. On the other hand, those remote requests will always have network latency anyway, so maybe it would be better to optimize the local case where it could actually matter, by keeping this not nullable. Eh, I doubt it makes a big difference either way.
___
*[`pkg/kv/kvserver/batcheval/cmd_scan.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUWNPy0oGCgYylc7LM:-NPUWNPy0oGCgYylc7LN:b-m3w2a1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/kv/kvserver/batcheval/cmd_scan.go#L76)):*
> ```Go
> 		}
> 		if len(scanRes.ColBatches) > 0 {
> 			reply.ColBatches.ColBatches = scanRes.ColBatches
> ```

So I guess this works because `rpc.internalClientAdapter` bypasses gRPC altogether, and just hands the `*kvpb.BatchResponse` to the kvclient, is that right? Pretty cool.
___
*[`pkg/sql/colfetcher/cfetcher_wrapper.go` line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUfuwYBmA0lJrC5L3q:-NPUfuwYBmA0lJrC5L3r:b-2e4uli) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colfetcher/cfetcher_wrapper.go#L116)):*
> ```Go
> 		prevBatchMemUsage := c.fetcherAcc.Used()
> 		if prevBatchMemUsage > 0 {
> 			if err := c.converterAcc.Grow(ctx, prevBatchMemUsage); err != nil {
> ```

When does this shrink?
___
*[`pkg/sql/colfetcher/cfetcher_wrapper.go` line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUalI38o-DyRC3Dg1Q:-NPUalI42JhgzlAAOngq:b-r992w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colfetcher/cfetcher_wrapper.go#L200)):*
> ```Go
> 		// serialize the response.
> 		for _, t := range tableArgs.typs {
> 			if t.Family() == types.EnumFamily {
> ```

question: From looking at the definition of `types.(*T).IsHydrated` I'm wondering, should this check (and other similar checks) be `t.UserDefined()` instead of `t.Family() == types.EnumFamily`?
___
*[`pkg/storage/col_mvcc.go` line 487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPURs6WBOlGxymIy_8A:-NPURk6pAkbLZ8kwFS1H:b-uymdot) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/storage/col_mvcc.go#L487)):*
> ```Go
> 			res.KVData = append(res.KVData, b)
> 		} else {
> 			res.ColBatches = append(res.ColBatches, colBatch)
> ```

Should we add some kind of assertion here that all previous results from `wrapper.NextBatch` in this loop were also unserialized? (Is this an invariant?)

**Review by DrewKimball — COMMENTED** [review]
> Reviewed 38 of 38 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95033)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @yuzefovich)

___
*[`pkg/sql/colexec/colexecutils/deserializer.go` line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPTWdt578QXk5r_76Tv:-NPTWdt578QXk5r_76Tw:bdoheml) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colexec/colexecutils/deserializer.go#L32)):*
> ```Go
> 
> 	batch     coldata.Batch
> 	data      []array.Data
> ```

Should we account for this?
___
*[`pkg/sql/colfetcher/cfetcher_wrapper.go` line 79 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPTXSWm1ncqQRq-ZD2z:-NPTXSWm1ncqQRq-ZD3-:b1jxxtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colfetcher/cfetcher_wrapper.go#L79)):*
> ```Go
> 	// only grows / shrinks the account according to its own usage and never
> 	// relies on the total used value.
> 	converterAcc *mon.BoundAccount
> ```

[nit] Could you rename this field since it isn't just used for the converter now?
___
*[`pkg/sql/colfetcher/cfetcher_wrapper.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPTYl04-8xqVfeuKOtc:-NPTYl04-8xqVfeuKOtd:beeoqx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colfetcher/cfetcher_wrapper.go#L114)):*
> ```Go
> 		// of the batch that was returned on the previous call to NextBatch
> 		// before we call nextBatchAdapter.
> 		prevBatchMemUsage := c.fetcherAcc.Used()
> ```

What if we passed a detached allocator to `cFetcher`, and then made it the responsibility of the `cFetcherWrapper` to account for every batch? Would that simplify the logic here? If serializing, we could just adjust by the difference with the previous batch's memory on each call to `fetcher.NextBatch`.
___
*[`pkg/sql/colfetcher/cfetcher_wrapper.go` line 264 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPTW-Wj6qkGRk-lOwyd:-NPTW-Wj6qkGRk-lOwye:b-2lz878) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colfetcher/cfetcher_wrapper.go#L264)):*
> ```Go
> ) ([]coldata.Batch, error) {
> 	// This memory monitor is not connected to the memory accounting system
> 	// since the accounting for these batches will be done by the SQL client.
> ```

This monitor/bound account is basically a no-op, right? Could we just pass a nil `BoundAccount` to `NewAllocator`?
___
*[`pkg/sql/colmem/allocator.go` line 720 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUuV8B-D77fu2RZtfx:-NPUuV8B-D77fu2RZtfy:bgupwj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colmem/allocator.go#L720)):*
> ```Go
> 	// ResetMaybeReallocate call. At the moment, it can only be set by the
> 	// SetAccountingHelper.
> 	noBatchReuse bool
> ```

[nit] Here and elsewhere, a name like `alwaysReallocate` feels a little more straightforward.
___
*[`pkg/sql/row/kv_batch_fetcher.go` line 547 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUt7N-6bxJ18zFjk5r:-NPUt7N-6bxJ18zFjk5s:b-cqrfrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/row/kv_batch_fetcher.go#L547)):*
> ```Go
> // under-accounting in most places, this seems acceptable.
> func popBatch(
> 	batches [][]byte, colBatches []coldata.Batch,
> ```

[nit] Seems a little nicer to just have a second function that does the same thing for colBatches, but up to you.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95033)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @michae2, and @sumeerbhola)

___
*[`pkg/kv/kvpb/api.proto` line 645 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUYXQODdj-1ZkhUuzr:-NPnlxH-1QVh-PhOklr7:b-4wk11t) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/kv/kvpb/api.proto#L645)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

useless blathering: On one hand, I would expect the majority of ScanRequests to cross the network and need serialization, suggesting that we should optimize for that more common remote case and set this nullable. On the other hand, those remote requests will always have network latency anyway, so maybe it would be better to optimize the local case where it could actually matter, by keeping this not nullable. Eh, I doubt it makes a big difference either way.
</blockquote></details>

Probably you're right that it's likely to not matter much. I'll keep this as non-nullable since then the code seems nicer (`batch_responses` and `col_batches` fields then look more similar).
___
*[`pkg/kv/kvserver/batcheval/cmd_scan.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUWNPy0oGCgYylc7LM:-NPngACi0ox3AkXFsxtg:b-9fssmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/kv/kvserver/batcheval/cmd_scan.go#L76)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

So I guess this works because `rpc.internalClientAdapter` bypasses gRPC altogether, and just hands the `*kvpb.BatchResponse` to the kvclient, is that right? Pretty cool.
</blockquote></details>

Yes, precisely.
___
*[`pkg/sql/colexec/colexecutils/deserializer.go` line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPTWdt578QXk5r_76Tv:-NPngLNy3jK64QSlWn4G:b-6m9qvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colexec/colexecutils/deserializer.go#L32)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should we account for this?
</blockquote></details>

This `[]array.Data` seems negligible - `array.Data` seems to be about 88 bytes in size, and we have one object per vector in the batch, with these objects being reused on each `Deserialize` call.
___
*[`pkg/sql/colfetcher/cfetcher_wrapper.go` line 79 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPTXSWm1ncqQRq-ZD2z:-NPngxAM56MbHg1cIUNR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colfetcher/cfetcher_wrapper.go#L79)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Could you rename this field since it isn't just used for the converter now?
</blockquote></details>

Done.
___
*[`pkg/sql/colfetcher/cfetcher_wrapper.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPTYl04-8xqVfeuKOtc:-NPnvP7i8PPEH3Y3meAA:b-86mr6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colfetcher/cfetcher_wrapper.go#L114)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

What if we passed a detached allocator to `cFetcher`, and then made it the responsibility of the `cFetcherWrapper` to account for every batch? Would that simplify the logic here? If serializing, we could just adjust by the difference with the previous batch's memory on each call to `fetcher.NextBatch`.
</blockquote></details>

I like it, done.
___
*[`pkg/sql/colfetcher/cfetcher_wrapper.go` line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUfuwYBmA0lJrC5L3q:-NPnh-8nCV0fY444GRi3:b-ckj6e9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colfetcher/cfetcher_wrapper.go#L116)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

When does this shrink?
</blockquote></details>

It doesn't until we're done processing the ScanRequest (when this will shrink when the account is closed). This is by design - all `coldata.Batch`es are live (they are accumulated into a slice), so we just move the accounting for each batch from `fetcherAccount` into `converterAccount`, but we do need to keep the accounting live until the very end.
___
*[`pkg/sql/colfetcher/cfetcher_wrapper.go` line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUalI38o-DyRC3Dg1Q:-NPnnzV3AlwDKp0D56gw:blbi8hw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colfetcher/cfetcher_wrapper.go#L200)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

question: From looking at the definition of `types.(*T).IsHydrated` I'm wondering, should this check (and other similar checks) be `t.UserDefined()` instead of `t.Family() == types.EnumFamily`?
</blockquote></details>

Enum type is the only user-defined type that we support (this is checked in `colbuilder/execplan.go`). This wasn't well documented, so I added some comments and an assertion.
___
*[`pkg/sql/colfetcher/cfetcher_wrapper.go` line 264 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPTW-Wj6qkGRk-lOwyd:-NPniDOg18ZMbqTvpqI2:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colfetcher/cfetcher_wrapper.go#L264)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

This monitor/bound account is basically a no-op, right? Could we just pass a nil `BoundAccount` to `NewAllocator`?
</blockquote></details>

Good point, done.
___
*[`pkg/sql/colmem/allocator.go` line 720 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUuV8B-D77fu2RZtfx:-NPnjrw0Bz4TjJsuSRe-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colmem/allocator.go#L720)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Here and elsewhere, a name like `alwaysReallocate` feels a little more straightforward.
</blockquote></details>

Done.
___
*[`pkg/sql/row/kv_batch_fetcher.go` line 547 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPUt7N-6bxJ18zFjk5r:-NPnk2rh96UgpfcXEHqz:bfhh834) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/row/kv_batch_fetcher.go#L547)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Seems a little nicer to just have a second function that does the same thing for colBatches, but up to you.
</blockquote></details>

This would result in more code overall (since we'd need to effectively push `if` out of this function outside, into the callsites), so I'll keep it as is.
___
*[`pkg/storage/col_mvcc.go` line 487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPURs6WBOlGxymIy_8A:-NPnlFOQ-4MrpoOjkjZY:blmel63) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/storage/col_mvcc.go#L487)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should we add some kind of assertion here that all previous results from `wrapper.NextBatch` in this loop were also unserialized? (Is this an invariant?)
</blockquote></details>

Good point, added assertions and adjusted the comments.

**Review by DrewKimball — APPROVED** [review]
:lgtm: Nice work!

> Reviewed 28 of 28 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-:-NPt5k-G13u-L_q8DhiU:bcmw7ss)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2, @sumeerbhola, and @yuzefovich)

___
*[`pkg/sql/colexec/colexecutils/deserializer.go` line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95033#-NPTWdt578QXk5r_76Tv:-NPszJO9-y1OOm6ntCd9:btzo0lz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0a8bf5603c167d8ded286ae2a7ea753a98c51a/pkg/sql/colexec/colexecutils/deserializer.go#L32)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This `[]array.Data` seems negligible - `array.Data` seems to be about 88 bytes in size, and we have one object per vector in the batch, with these objects being reused on each `Deserialize` call.
</blockquote></details>

I see, didn't realize `array.Data` is small. Sounds good to me.

**Comment by michae2** [comment]
I haven't had a chance to look again yet, but I don't think you need to wait for me. Please feel free to merge!

**Comment by yuzefovich** [comment]
Ack, thanks for the reviews!

bors r+

**Comment by yuzefovich** [comment]
This flake seems legitimate.

bors r+

**Comment by yuzefovich** [comment]
I made a minor change to include the start key as the error detail (`TestMemoryLimit` was failing because it didn't see expected error prefix).

bors r+
